File: cses/02_sorting_searching/1642_sum-of-four-values/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    n = int(data[0])
    x = int(data[1])
    a = [int(v) for v in data[2:2 + n]]

    # hashmap: tổng cặp -> (k, l) với k < l, cả hai chỉ số đều đã cố định
    seen = {}
    get = seen.get

    for i in range(n):
        ai = a[i]
        # TÌM: coi i là chỉ số nhỏ của cặp cao; với mọi j > i tìm need trong seen
        for j in range(i + 1, n):
            p = get(x - ai - a[j])
            if p is not None:
                k, l = p
                sys.stdout.write("%d %d %d %d\n" % (k + 1, l + 1, i + 1, j + 1))
                return
        # THÊM: mọi cặp (k, i) với k < i (chỉ giữ cặp đầu tiên cho mỗi tổng),
        # thực hiện SAU bước tìm nên mọi cặp trong seen đều có chỉ số < i
        for k in range(i):
            s = a[k] + ai
            if s not in seen:
                seen[s] = (k, i)

    sys.stdout.write("IMPOSSIBLE\n")


main()

File: cses/02_sorting_searching/1642_sum-of-four-values/solution.cpp
#include <iostream>
#include <unordered_map>
#include <utility>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n;
    long long target;
    std::cin >> n >> target;

    std::vector<long long> values(n);
    for (long long& value : values) {
        std::cin >> value;
    }

    // hashmap ánh xạ: tổng cặp -> một cặp chỉ số (k, l) có tổng đó
    std::unordered_map<long long, std::pair<int, int>> seen;
    seen.max_load_factor(0.7F);
    seen.reserve(static_cast<std::size_t>(n) * static_cast<std::size_t>(n) / 2);

    for (int i = 0; i < n; ++i) {
        // TÌM: coi i là chỉ số nhỏ của cặp cao; mọi cặp trong seen đều kết thúc
        // trước i nên bốn chỉ số luôn phân biệt
        for (int j = i + 1; j < n; ++j) {
            const long long needed = target - values[i] - values[j];
            const auto it = seen.find(needed);
            if (it != seen.end()) {
                std::cout << it->second.first + 1 << ' '
                          << it->second.second + 1 << ' '
                          << i + 1 << ' ' << j + 1 << '\n';
                return 0;
            }
        }

        // THÊM: chỉ thêm sau khi đã tìm xong, giữ ràng buộc k < i < cặp cao;
        // try_emplace bảo đảm chỉ giữ cặp đầu tiên cho mỗi tổng
        for (int k = 0; k < i; ++k) {
            seen.try_emplace(values[k] + values[i], k, i);
        }
    }

    std::cout << "IMPOSSIBLE\n";
    return 0;
}
